File: haskell/Spark.hs
module Spark where

import Data.List
import Data.Maybe
import System.Random
import System.Random.Shuffle

type Partition a = [a]

type RDD a = [Partition a]

shuffleR :: RandomGen g => g -> [a] -> ([a], g)
shuffleR g [] = ([], g)
shuffleR g xs =
  let loop acc 0 g = (reverse acc, g)
      loop acc n g = let (r, g') = randomR (0, n) g
                     in loop (r:acc) (pred n) g'
      make_rs n g = loop [] n g
      (rs, g') = make_rs (length xs - 1) g 
  in (shuffle xs rs, g')

mapR :: RandomGen g => g -> (a -> b) -> [a] -> ([b], g)
mapR g f xs = shuffleR g (map f xs)

concatMapR :: RandomGen g => g -> (a -> [b]) -> [a] -> ([b], g)
concatMapR g f xs = (concat yss, g')
                    where (yss, g') = mapR g f xs

reducel :: (a -> a -> a) -> [a] -> a
reducel f (x:xs) = foldl f x xs

aggregate :: RandomGen g => 
             g -> b -> (b -> a -> b) -> (b -> b -> b) -> RDD a -> (b, g)
aggregate g z seq comb rdd =
    let (presults, g') = mapR g (foldl seq z) rdd in
    (foldl comb z presults, g')

reduce :: RandomGen g => g -> (a -> a -> a) -> RDD a -> (a, g)
reduce g comb rdd =
    let (presults, g') = mapR g (reducel comb) rdd in
    (reducel comb presults, g')

splitR :: RandomGen g =>
          g -> [a] -> ([a], [a], g)
splitR g xs =
  let (cut, g') = randomR (0, length xs) g
      (ls, rs) = splitAt cut xs
  in (ls, rs, g')

repartitionR :: RandomGen g => g -> [a] -> ([[a]], g)
repartitionR g [] = ([], g)
repartitionR g xs =
  let (ys, g') = shuffleR g xs
      f x (cur, g, acc) =
        let (cut, g') = random g
        in if cut then ([x], g', cur:acc)
            else (x:cur, g', acc)
      (p, g'', ps) = foldr f ([last ys], g', []) 
                      (reverse (tail (reverse ys)))
  in (p:ps, g'')

divideR :: RandomGen g =>
           g -> [a] -> ([a], a, a, [a], g)
divideR g xs =
  let (cut, g') = randomR (0, length xs - 2) g 
      (ls, l:r:rs) = splitAt cut xs
  in (ls, l, r, rs, g')

applyR :: RandomGen g =>
          g -> (b -> b -> b) -> Partition b -> (b, g)
applyR g comb [r] = (r, g)
applyR g comb [r, r'] = (comb r r', g)
applyR g comb rs = 
  let (ls', l', r', rs', g') = divideR g rs
  in applyR g' comb (ls' ++ [comb l' r'] ++ rs')

treeAggregate :: RandomGen g =>
                 g -> b -> (b -> a -> b) -> (b -> b -> b) -> RDD a -> (b, g)
treeAggregate g z seq comb rdd =
  let (presults, g') = mapR g (foldl seq z) rdd
  in applyR g' comb presults

treeReduce :: RandomGen g =>
              g -> (a -> a -> a) -> RDD a -> (a, g)
treeReduce g comb rdd =
  let (presults, g') = mapR g (reducel comb) rdd
  in applyR g' comb presults

type PairRDD a b = RDD (a, b)

hasKey :: Eq a => a -> Partition (a, b) -> Bool
hasKey key ps = case lookup key ps of
                  Just _  -> True   
                  Nothing -> False

hasValue :: Eq a => a -> b -> Partition (a, b) -> b
hasValue key val ps = case lookup key ps of
                        Just v -> v
                        Nothing -> val

addTo :: Eq a => a -> b -> Partition (a, b) -> Partition (a, b)
addTo key val ps = foldl (\r (k, v) -> if key == k then r else (k, v):r) 
                         [(key, val)] ps

aggregateByKey :: (Eq a, RandomGen g) => 
                  g -> c -> (c -> b -> c) -> (c -> c -> c) ->
                  PairRDD a b -> (PairRDD a c, g)
aggregateByKey g z mergeComb mergeValue pairRdd =
    let mergeBy f r (k, v) = 
            addTo k (f (hasValue k z r) v) r
        (preAgg, g') = 
            concatMapR g (foldl (mergeBy mergeComb) []) pairRdd
    in repartitionR g' (foldl (mergeBy mergeValue) [] preAgg)

aggregateWithKey :: (Eq a, RandomGen g) => g -> a -> c -> (c -> b -> c) -> 
                    (c -> c -> c) -> PairRDD a b -> (c, g)
aggregateWithKey g key z mergeComb mergeValue pairRdd =
  let select (k, _) = key == k
      vrdd = filter (not . null) 
             (map ((map snd) . (filter select)) pairRdd) in
  aggregate g z mergeComb mergeValue vrdd

reduceByKey :: (Eq a, RandomGen g) => 
               g -> (b -> b -> b) -> PairRDD a b -> (PairRDD a b, g)
reduceByKey g mergeValue pairRdd =
    let merge r (k, v) =
          case lookup k r of
            Just v' -> addTo k (mergeValue v' v) r
            Nothing -> addTo k v r
        (preAggrs, g') =
            concatMapR g (foldl merge []) pairRdd in
    repartitionR g' (foldl merge [] preAggrs)

reduceWithKey :: (Eq a, RandomGen g) => 
                 g -> a -> (b -> b -> b) -> PairRDD a b -> (b, g)
reduceWithKey g key mergeValue pairRdd =
  let select (k, _) = key == k
      vrdd = filter (not . null)
             (map ((map snd) . (filter select)) pairRdd) in
  reduce g mergeValue vrdd

type VertexId = Int
type VertexRDD a = PairRDD VertexId a
type EdgeRDD b = RDD (VertexId, VertexId, b)
data GraphRDD a b = Graph { vertexRdd :: VertexRDD a,
                            edgeRdd :: EdgeRDD b }

aggregateMessagesWithActiveVertices :: RandomGen g =>
  g -> (VertexId -> a -> VertexId -> a -> b -> [(VertexId, c)]) ->
  (c -> c -> c) -> [VertexId] -> GraphRDD a b -> 
  (VertexRDD c, g)
aggregateMessagesWithActiveVertices g sendMsg mergeMsg active graphRdd =
  let 
    isActive (srcId, dstId, _) = 
      srcId `elem` active || dstId `elem` active
    vAttrs = concat (vertexRdd graphRdd)
    f edge =
      if isActive edge then
        let (srcId, dstId, edgeAttr) = edge
            srcAttr = fromJust (lookup srcId vAttrs)
            dstAttr = fromJust (lookup dstId vAttrs) in
        sendMsg srcId srcAttr dstId dstAttr edgeAttr
      else []
    pairRdd = map (concatMap f) (edgeRdd graphRdd)
  in reduceByKey g mergeMsg pairRdd

aggregateMessages :: RandomGen g =>
  g -> (VertexId -> a -> VertexId -> a -> b -> [(VertexId, c)]) ->
  (c -> c -> c) -> GraphRDD a b -> (VertexRDD c, g)
aggregateMessages g sendMsg mergeMsg graphRdd =
  let vertices = concatMap (map fst) (vertexRdd graphRdd)
  in aggregateMessagesWithActiveVertices g
       sendMsg mergeMsg vertices graphRdd

mapVertexRDD :: (VertexId -> a -> b) -> VertexRDD a -> VertexRDD b
mapVertexRDD f vRdd = 
  map (map (\(i, attr) -> (i, f i attr))) vRdd

mapVertices :: (VertexId -> a -> c) -> GraphRDD a b -> GraphRDD c b
mapVertices updater gRdd =
  Graph {
    vertexRdd = mapVertexRDD updater (vertexRdd gRdd),
    edgeRdd = edgeRdd gRdd }

joinGraph :: (VertexId -> a -> c -> a) 
             -> GraphRDD a b -> VertexRDD c -> GraphRDD a b
joinGraph joiner gRdd vRdd =
  let assoc = concat vRdd
      updt i attr =
        case lookup i assoc of
          Just v  -> joiner i attr v
          Nothing -> attr
  in mapVertices updt gRdd

pregel :: RandomGen g => g -> c -> (VertexId -> a -> c -> a) ->
 (VertexId -> a -> VertexId -> a -> b -> [(VertexId, c)])
 -> (c -> c -> c) -> GraphRDD a b -> (GraphRDD a b, g)
pregel g initMsg vprog sendMsg mergeMsg graphRdd =
  let
    initG = let f i attr = vprog i attr initMsg in
      mapVertices f graphRdd
    (initVRdd, g') = aggregateMessages g sendMsg mergeMsg initG
    loop g curG [] = (curG, g)
    loop g curG vRdd =
      let
        curG' = joinGraph vprog curG vRdd
        active = concatMap (map fst) vRdd
        (vRdd', g') = aggregateMessagesWithActiveVertices g
                         sendMsg mergeMsg active curG'
      in loop g' curG' vRdd'
  in loop g' initG initVRdd

File: scala/Spark.scala
import scala.language.implicitConversions

object Prelude {

  type Partition[A] = List[A]
  type RDD[A] = List[Partition[A]]
  type PairRDD[A, B] = RDD[Pair[A, B]]

  class Pair[A, B] (val key: A, val value: B)

  implicit def TupleToPair[A, B] (t: (A, B)): Pair[A, B] = new Pair(t._1, t._2)

  /** ************************
    * Deterministic operations
    * ************************/

  def map[A, B] (pa: Partition[A])(fn: A => B): Partition[B] =
    pa match {
      case Nil     => Nil
      case x :: xs => fn(x) :: map(xs)(fn)
    }

  def filter[A] (pa: Partition[A])(fn: A => Boolean): Partition[A] =
    pa match {
      case Nil     => Nil
      case x :: xs => if (fn(x)) x :: filter(xs)(fn) else filter(xs)(fn)
    }

  def foldl[A, B] (pa: Partition[A])(z: B)(fn: (B, A) => B): B =
    pa match {
      case Nil     => z
      case x :: xs => foldl(xs)(fn(z, x))(fn)
    }

  def reducel[A] (pa: Partition[A])(fn: (A, A) => A): A =
    pa match {
      case x :: xs => foldl(xs)(x)(fn)
      case Nil     => throw new Error("empty list")
    }

  def concat[A] (rdd: RDD[A]): Partition[A] =
    rdd match {
      case Nil     => Nil
      case x :: xs => x ++ concat(xs)
    }

  def concatMap[A, B] (pa: Partition[A])(fn: A => Partition[B]): Partition[B] =
    concat(map(pa)(fn))

  def hasValue[A, B] (pa: Partition[Pair[A, B]])(default: B)(key: A) =
    foldl(pa)(default) {
      (value, pair) => if (pair.key == key) pair.value else value
    }

  def hasKey[A, B] (pa: Partition[Pair[A, B]])(key: A): Boolean =
    foldl(pa)(false) {
      (found, pair) => if (pair.key == key) true else found
    }

  def addTo[A, B] (pa: Partition[Pair[A, B]])(pair: Pair[A, B]): Partition[Pair[A, B]] =
    foldl(pa)(List(pair)) {
      (p1, p2) => if (pair.key == p2.key) p1 else p2 :: p1
    }

  def lookup[A, B] (pa: Partition[Pair[A, B]])(key: A): Option[B] =
    foldl(pa)(None: Option[B]) {
      (p1, p2) => if (p2.key == key) Some(p2.value) else p1
    }

  /** ****************************
    * Non-deterministic operations
    * ****************************/

  /* Nondeterministic map */
  def mapX[A, B] (pa: Partition[A])(fn: A => B): Partition[B] = map(pa)(fn)

  /* Nondeterministic concatMap */
  def concatMapX[A, B] (pa: Partition[A])(fn: A => Partition[B]): Partition[B] = concat(mapX(pa)(fn))

  /* Nondeterministic re-partitioning */
  def shufflePartitionX[A] (pa: Partition[A]): RDD[A] = List(pa)

  def aggregate[A, B] (rdd: RDD[A])(z: B)(seq: (B, A) => B)(comb: (B, B) => B): B =
    foldl(mapX(rdd)(foldl(_)(z)(seq)))(z)(comb)

  def reduce[A] (rdd: RDD[A])(comb: (A, A) => A): A =
    reducel(mapX(rdd)(reducel(_)(comb)))(comb)

  def aggregateByKey[A, B, C] (rdd: PairRDD[A, B])(z: C)(mergeComb: (C, B) => C, mergeValue: (C, C) => C): PairRDD[A, C] = {
    def seq[D] (f: (C, D) => C)(pa: Partition[Pair[A, C]], p: Pair[A, D]) =
      addTo(pa)(p.key, f(hasValue(pa)(z)(p.key), p.value))
    val zero: Partition[Pair[A, C]] = Nil
    val preAggrs = concatMapX(rdd)(foldl(_)(zero)(seq(mergeComb)))
    shufflePartitionX(foldl(preAggrs)(zero)(seq(mergeValue)))
  }

  def aggregateWithKey[A, B, C] (rdd: PairRDD[A, B])(key: A)(z: C)(mergeComb: (C, B) => C, mergeValue: (C, C) => C): C = {
    val preAggrs = filter(map(rdd) {
      pa => map(filter(pa)(_.key == key))(_.value)
    })(_ != Nil)
    aggregate(preAggrs)(z)(mergeComb)(mergeValue)
  }

  def reduceByKey[A, B] (rdd: PairRDD[A, B])(comb: (B, B) => B): PairRDD[A, B] = {
    def merge (pa: Partition[Pair[A, B]], p: Pair[A, B]) =
      lookup(pa)(p.key) match {
        case None    => addTo(pa)(p.key, p.value)
        case Some(v) => addTo(pa)(p.key, comb(v, p.value))
      }
    val z: Partition[Pair[A, B]] = Nil
    shufflePartitionX(foldl(concatMapX(rdd)(foldl(_)(z)(merge)))(z)(merge))
  }

  def reduceWithKey[A, B] (rdd: PairRDD[A, B])(key: A)(comb: (B, B) => B): B = {
    val preAggrs = map(rdd) {
      pa => map(filter(pa)(_.key == key))(_.value)
    }
    reduce(filter(preAggrs)(_ != Nil))(comb)
  }
}

object GraphX {

  import Prelude._

  case class Edge[B] (srcId: VertexId,
                      dstId: VertexId,
                      attr: B) extends Pair((srcId, dstId), attr)

  case class Vertex[A] (id: VertexId,
                        attr: A) extends Pair(id, attr)

  case class EdgeTriplet[A, B] (srcId: VertexId, srcAttr: A,
                                dstId: VertexId, dstAttr: A,
                                attr: B) extends Pair((srcId, dstId), attr)

  type VertexId = Int
  type VertexRDD[A] = RDD[Vertex[A]]
  type EdgeRDD[B] = RDD[Edge[B]]

  case class GraphRDD[A, B] (vertexRDD: VertexRDD[A], edgeRDD: EdgeRDD[B])

  implicit def PairRDDToVertexRDD[A] (rdd: PairRDD[VertexId, A]): VertexRDD[A] =
    map(rdd)(map(_)(p => Vertex(p.key, p.value)))

  def aggregateMessages[A, B, C] (graph: GraphRDD[A, B])
                                 (sendMsg: EdgeTriplet[A, B] => List[Vertex[C]])
                                 (mergeMsg: (C, C) => C): VertexRDD[C] =
    aggregateMessagesWithActiveSet(graph)(sendMsg)(mergeMsg)(concat(graph.vertexRDD))

  def aggregateMessagesWithActiveSet[A, B, C] (graph: GraphRDD[A, B])
                                              (sendMsg: EdgeTriplet[A, B] => List[Vertex[C]])
                                              (mergeMsg: (C, C) => C)
                                              (activeSet: List[Vertex[A]]): VertexRDD[C] = {
    val edgeRDD = filter(map(graph.edgeRDD)(filter(_) {
      e => hasKey(activeSet)(e.srcId) || hasKey(activeSet)(e.dstId)
    }))(_ != Nil)
    val vAttrs = concat(graph.vertexRDD)
    def f (edge: Edge[B]) = {
      val Some(srcAttr) = lookup(vAttrs)(edge.srcId)
      val Some(dstAttr) = lookup(vAttrs)(edge.dstId)
      sendMsg(EdgeTriplet(edge.srcId, srcAttr, edge.dstId, dstAttr, edge.attr))
    }
    reduceByKey(mapX(edgeRDD)(concatMap(_)(f)))(mergeMsg)
  }

  def mapVertices[A, C] (vertices: VertexRDD[A])(fn: Vertex[A] => C): VertexRDD[C] =
    mapX(vertices)(map(_)(v => Vertex(v.id, fn(v))))

  def Pregel[A, B, C] (graph: GraphRDD[A, B])
                      (initMsg: C)
                      (vprog: (Vertex[A], C) => A)
                      (sendMsg: EdgeTriplet[A, B] => List[Vertex[C]])
                      (mergeMsg: (C, C) => C): GraphRDD[A, B] = {

    val initGraph = GraphRDD(mapVertices(graph.vertexRDD)(vprog(_, initMsg)), graph.edgeRDD)
    val initMsgs = concat(aggregateMessages(initGraph)(sendMsg)(mergeMsg))

    def loop (graph: GraphRDD[A, B], msgs: List[Vertex[C]]): GraphRDD[A, B] =
      if (msgs == Nil)
        graph
      else {
        val newGraph = GraphRDD(
          mapVertices(graph.vertexRDD) {
            v => lookup(msgs)(v.id) match {
              case None       => v.attr
              case Some(attr) => vprog(v, attr)
            }
          },
          graph.edgeRDD
        )
        val activeSet = filter(concat(graph.vertexRDD))(v => hasKey(msgs)(v.id))
        val newMsgs = concat(aggregateMessagesWithActiveSet(newGraph)(sendMsg)(mergeMsg)(activeSet))
        loop(newGraph, newMsgs)
      }
    loop(initGraph, initMsgs)
  }
}
